### DISCUSSION TITLE
When testing, `tcp connect error` occurs.

### DISCUSSION THREAD

**Discussion by AkifumiSato** [hyperium/hyper#2756] [Q&A] [answered]
**Version**
0.14.16

**Platform**
Darwin CA8103.local 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64

**Description**
I wrote a test using the `Server` and `Server::bind` exported from axum, but the error shown in the title occurs.

```rust
let addr = SocketAddr::from(([127, 0, 0, 1], 3000));

tokio::spawn(async move {
    axum::Server::bind(&addr)
        .serve(create_route().into_make_service())
        .await
        .expect("unexpected server error");
});

let client = hyper::Client::new();

let response = client
    .request(
        Request::builder()
            .uri(format!("http://{}", addr))
            .body(Body::empty())
            .unwrap(),
    )
    .await
    .unwrap();
```

```
failures:

---- test::should_return_hello_world stdout ----
thread 'test::should_return_hello_world' panicked at 'called `Result::unwrap()` on an `Err` value: hyper::Error(Connect, ConnectError("tcp connect error", Os { code: 61, kind: ConnectionRefused, message: "Connection refused" }))', src/main.rs:97:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I rewrote it using `Server::from_tcp` and the test passed successfully.

```rust
let listener = TcpListener::bind("127.0.0.1:8000".parse::<SocketAddr>().unwrap()).unwrap();
let addr = listener.local_addr().unwrap();

...

tokio::spawn(async move {
    axum::Server::from_tcp(listener)
    .unwrap()
    .serve(create_route().into_make_service())
    .await
    .expect("unexpected server error");
});
```

Both worked fine when I did a `cargo run`.
Is the difference in behavior between `bind` and `from_tcp` a bug?

**Comment by olix0r** [comment, marked as answer]
This is probably a race. In the first version, the client connection can occur before the spawned tasks actually binds the server. In the second example, you've bound the server before spawning the server task.

**Reply by seanmonstar** [reply]
Exactly. The `Server::bind` happens in a spawned task, which doesn't necessarily execute immediately, so the client tried to connect to an address that wasn't bound yet.

**Reply by AkifumiSato** [reply]
Thanks a lot!
